### ISSUE TITLE
OneDrive offline access fails to fallback to cached file

### ISSUE THREAD

**Issue by jbujalance** [PhilippC/keepass2android#941]
Hi,
First of all, thanks for the effort put into this app, it is great.

I am experiencing what I think could be a bug when using the application offline.
My database is synchronized in OneDrive, and when I open the Keepass2Android application offline, I get an infinite loading icon like this:
![imagen](https://user-images.githubusercontent.com/30215288/64067251-e3402980-cc25-11e9-90e4-2da709c2f9d0.png)
Here is an extract of the application log when I reproduce the problem. The error occurs at 19:20:21:489:
```
31/08/2019 19:15:36:905 -- AppSettingsActivity.OnPause
31/08/2019 19:15:37:803 -- AppSettingsActivity.OnStop
31/08/2019 19:15:53:393 -- AppSettingsActivity.OnStart
31/08/2019 19:15:53:396 -- AppSettingsActivity.OnResume
31/08/2019 19:15:53:397 --  DB null
31/08/2019 19:15:54:485 -- AppSettingsActivity.OnPause
31/08/2019 19:15:54:557 -- PasswordActivity.OnStart
31/08/2019 19:15:54:560 -- PasswordActivity.OnResume
31/08/2019 19:15:54:561 --  DB null
31/08/2019 19:15:54:563 -- starting: True, Finishing: False, _performingLoad: False
31/08/2019 19:15:55:378 -- AppSettingsActivity.OnStop
31/08/2019 19:15:55:383 -- AppSettingsActivity.OnDestroyTrue
31/08/2019 19:15:57:544 -- PasswordActivity.OnPause
31/08/2019 19:15:57:574 -- StackBaseActivity.OnActivityResult Canceled/1
31/08/2019 19:15:57:577 -- TryGetFromActivityResult: no data
31/08/2019 19:15:58:13 -- PasswordActivity.OnStop
31/08/2019 19:15:58:24 -- PasswordActivity.OnDestroyTrue
31/08/2019 19:16:24:49 -- Creating application keepass2android.keepass2android. Version=138
31/08/2019 19:16:25:231 -- KeePass.OnCreate
31/08/2019 19:16:25:238 -- KeePass:apptask=
31/08/2019 19:16:25:250 -- Loaded task keepass2android.NullTask
31/08/2019 19:16:25:251 -- KeePass.OnCreate
31/08/2019 19:16:25:267 -- KeePass.OnStart
31/08/2019 19:16:25:268 -- KeePass.OnStart
31/08/2019 19:16:25:400 -- Loaded task keepass2android.NullTask
31/08/2019 19:16:25:510 -- FileSelect.OnCreate
31/08/2019 19:16:25:767 -- FileSelect.OnStart
31/08/2019 19:16:26:161 -- FileSelect.OnResume
31/08/2019 19:16:26:169 -- FileSelect.OnPause
31/08/2019 19:16:26:222 -- FSSA.OnCreate
31/08/2019 19:16:26:459 -- FileSelect.OnStop
31/08/2019 19:16:26:471 -- KeePass.OnStop
31/08/2019 19:16:26:480 -- KeePass.OnDestroyTrue
31/08/2019 19:16:26:486 -- KeePass.OnDestroyTrue
31/08/2019 19:16:55:760 -- Creating application keepass2android.keepass2android. Version=138
31/08/2019 19:16:55:946 -- KeePass.OnCreate
31/08/2019 19:16:55:953 -- KeePass:apptask=
31/08/2019 19:16:55:965 -- Loaded task keepass2android.NullTask
31/08/2019 19:16:55:966 -- KeePass.OnCreate
31/08/2019 19:16:55:982 -- KeePass.OnStart
31/08/2019 19:16:55:982 -- KeePass.OnStart
31/08/2019 19:16:56:120 -- Loaded task keepass2android.NullTask
31/08/2019 19:16:56:266 -- FileSelect.OnCreate
31/08/2019 19:16:56:573 -- FileSelect.OnStart
31/08/2019 19:16:58:330 -- PasswordActivity.OnCreate
31/08/2019 19:16:58:332 -- PasswordActivity:apptask=
31/08/2019 19:16:59:159 -- PasswordActivity.OnStart
31/08/2019 19:16:59:192 -- PasswordActivity.OnResume
31/08/2019 19:16:59:192 --  DB null
31/08/2019 19:16:59:197 -- starting: True, Finishing: False, _performingLoad: False
31/08/2019 19:16:59:301 -- PasswordActivity.OnPause
31/08/2019 19:16:59:327 -- Pre-loading database file starting
31/08/2019 19:16:59:366 -- CFS: OpenWhenNoLocalChanges
31/08/2019 19:16:59:368 -- CFS: hashing cached version
31/08/2019 19:16:59:642 -- PasswordActivity.OnResume
31/08/2019 19:16:59:643 --  DB null
31/08/2019 19:16:59:645 -- starting: False, Finishing: False, _performingLoad: False
31/08/2019 19:16:59:790 -- FileSelect.OnStop
31/08/2019 19:16:59:894 -- FileSelect.OnDestroyTrue
31/08/2019 19:16:59:901 -- KeePass.OnStop
31/08/2019 19:16:59:904 -- KeePass.OnDestroyTrue
31/08/2019 19:16:59:906 -- KeePass.OnDestroyTrue
31/08/2019 19:17:00:482 -- CFS: Files in Sync
31/08/2019 19:17:00:510 -- Pre-loading database file completed
31/08/2019 19:17:05:470 -- PasswordActivity.OnPause
31/08/2019 19:17:05:542 -- AppSettingsActivity.OnCreate
31/08/2019 19:17:05:544 -- AppSettingsActivity:apptask=
31/08/2019 19:17:05:840 -- AppSettingsActivity.OnStart
31/08/2019 19:17:05:845 -- AppSettingsActivity.OnResume
31/08/2019 19:17:05:846 --  DB null
31/08/2019 19:17:06:321 -- PasswordActivity.OnStop
31/08/2019 19:17:10:538 -- AppSettingsActivity.OnPause
31/08/2019 19:17:19:809 -- AppSettingsActivity.OnStop
31/08/2019 19:17:32:37 -- AppSettingsActivity.OnStart
31/08/2019 19:17:32:43 -- AppSettingsActivity.OnResume
31/08/2019 19:17:32:44 --  DB null
31/08/2019 19:17:40:949 -- AppSettingsActivity.OnPause
31/08/2019 19:17:41:15 -- PasswordActivity.OnStart
31/08/2019 19:17:41:18 -- PasswordActivity.OnResume
31/08/2019 19:17:41:18 --  DB null
31/08/2019 19:17:41:20 -- starting: True, Finishing: False, _performingLoad: False
31/08/2019 19:17:41:859 -- AppSettingsActivity.OnStop
31/08/2019 19:17:41:864 -- AppSettingsActivity.OnDestroyTrue
31/08/2019 19:17:43:335 -- PasswordActivity.OnPause
31/08/2019 19:17:43:368 -- StackBaseActivity.OnActivityResult Canceled/1
31/08/2019 19:17:43:372 -- TryGetFromActivityResult: no data
31/08/2019 19:17:43:791 -- PasswordActivity.OnStop
31/08/2019 19:17:43:797 -- PasswordActivity.OnDestroyTrue
31/08/2019 19:20:19:194 -- KeePass.OnCreate
31/08/2019 19:20:19:197 -- KeePass:apptask=
31/08/2019 19:20:19:199 -- KeePass.OnCreate
31/08/2019 19:20:19:211 -- KeePass.OnStart
31/08/2019 19:20:19:212 -- KeePass.OnStart
31/08/2019 19:20:19:271 -- Loaded task keepass2android.NullTask
31/08/2019 19:20:19:318 -- FileSelect.OnCreate
31/08/2019 19:20:19:384 -- FileSelect.OnStart
31/08/2019 19:20:19:433 -- PasswordActivity.OnCreate
31/08/2019 19:20:19:435 -- PasswordActivity:apptask=
31/08/2019 19:20:21:426 -- PasswordActivity.OnStart
31/08/2019 19:20:21:433 -- PasswordActivity.OnResume
31/08/2019 19:20:21:434 --  DB null
31/08/2019 19:20:21:437 -- starting: True, Finishing: False, _performingLoad: False
31/08/2019 19:20:21:446 -- Pre-loading database file starting
31/08/2019 19:20:21:452 -- CFS: OpenWhenNoLocalChanges
31/08/2019 19:20:21:454 -- CFS: hashing cached version
31/08/2019 19:20:21:465 -- PasswordActivity.OnPause
31/08/2019 19:20:21:489 -- Error during http request
31/08/2019 19:20:21:604 -- System.Exception: Error during http request ---> Java.Lang.RuntimeException: Error during http request ---> Java.IO.IOException: Unable to resolve host "api.onedrive.com": No address associated with hostname ---> Java.Lang.RuntimeException: android_getaddrinfo failed: EAI_NODATA (No address associated with hostname)
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <a2068f8b3f6f40d0ad70bc362d4b0524>:0
  at Java.Interop.JniEnvironment+InstanceMethods.CallObjectMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x00069] in <daf82fa226e64866b6883f2960c1a0f1>:0
  at Android.Runtime.JNIEnv.CallObjectMethod (System.IntPtr jobject, System.IntPtr jmethod, Android.Runtime.JValue* parms) [0x0000e] in <f53e43de1d6c43078eb67fc63a0181e1>:0
  at Keepass2android.Javafilestorage.OneDriveStorage.OpenFileForRead (System.String p0) [0x00062] in <bbfdbb1fb19a4f0a8a8c95bc1465a6cb>:0
  at keepass2android.Io.JavaFileStorage.OpenFileForRead (KeePassLib.Serialization.IOConnectionInfo ioc) [0x0000d] in <d9555282dea14061ab4400355adad9e4>:0
   --- End of inner exception stack trace ---
  at keepass2android.Io.JavaFileStorage.OpenFileForRead (KeePassLib.Serialization.IOConnectionInfo ioc) [0x0002b] in <d9555282dea14061ab4400355adad9e4>:0
  at keepass2android.Io.OfflineSwitchableFileStorage.OpenFileForRead (KeePassLib.Serialization.IOConnectionInfo ioc) [0x00006] in <d9555282dea14061ab4400355adad9e4>:0
  at keepass2android.Io.CachingFileStorage.UpdateCacheFromRemote (KeePassLib.Serialization.IOConnectionInfo ioc, System.String cachedFilePath) [0x00000] in <d9555282dea14061ab4400355adad9e4>:0
  at keepass2android.Io.CachingFileStorage.OpenFileForReadWhenNoLocalChanges (KeePassLib.Serialization.IOConnectionInfo ioc, System.String cachedFilePath) [0x00023] in <d9555282dea14061ab4400355adad9e4>:0
  at keepass2android.Io.CachingFileStorage.OpenFileForRead (KeePassLib.Serialization.IOConnectionInfo ioc) [0x00030] in <d9555282dea14061ab4400355adad9e4>:0
31/08/2019 19:20:21:616 -- Pre-loading database file completed
31/08/2019 19:20:21:749 -- PasswordActivity.OnResume
31/08/2019 19:20:21:750 --  DB null
31/08/2019 19:20:21:751 -- starting: False, Finishing: False, _performingLoad: False
31/08/2019 19:20:22:11 -- FileSelect.OnStop
31/08/2019 19:20:22:84 -- FileSelect.OnDestroyTrue
31/08/2019 19:20:22:89 -- KeePass.OnStop
31/08/2019 19:20:22:90 -- KeePass.OnDestroyTrue
31/08/2019 19:20:22:91 -- KeePass.OnDestroyTrue
31/08/2019 19:20:26:751 -- PasswordActivity.OnPause
31/08/2019 19:20:26:816 -- PasswordActivity.OnStop
31/08/2019 19:20:35:708 -- Creating application keepass2android.keepass2android. Version=138
31/08/2019 19:20:35:888 -- KeePass.OnCreate
31/08/2019 19:20:35:895 -- KeePass:apptask=
31/08/2019 19:20:35:907 -- Loaded task keepass2android.NullTask
31/08/2019 19:20:35:908 -- KeePass.OnCreate
31/08/2019 19:20:35:924 -- KeePass.OnStart
31/08/2019 19:20:35:924 -- KeePass.OnStart
31/08/2019 19:20:36:58 -- Loaded task keepass2android.NullTask
31/08/2019 19:20:36:194 -- FileSelect.OnCreate
31/08/2019 19:20:36:455 -- FileSelect.OnStart
31/08/2019 19:20:36:845 -- FileSelect.OnResume
31/08/2019 19:20:36:855 -- FileSelect.OnPause
31/08/2019 19:20:36:912 -- FSSA.OnCreate
31/08/2019 19:20:37:125 -- FileSelect.OnStop
31/08/2019 19:20:37:140 -- KeePass.OnStop
31/08/2019 19:20:37:149 -- KeePass.OnDestroyTrue
31/08/2019 19:20:37:153 -- KeePass.OnDestroyTrue
31/08/2019 19:22:44:736 -- FileSelect.OnStart
31/08/2019 19:23:30:422 -- Creating application keepass2android.keepass2android. Version=138
31/08/2019 19:23:30:603 -- KeePass.OnCreate
31/08/2019 19:23:30:609 -- KeePass:apptask=
31/08/2019 19:23:30:621 -- Loaded task keepass2android.NullTask
31/08/2019 19:23:30:622 -- KeePass.OnCreate
31/08/2019 19:23:30:639 -- KeePass.OnStart
31/08/2019 19:23:30:639 -- KeePass.OnStart
31/08/2019 19:23:30:769 -- Loaded task keepass2android.NullTask
31/08/2019 19:23:30:879 -- FileSelect.OnCreate
31/08/2019 19:23:31:144 -- FileSelect.OnStart
31/08/2019 19:23:33:313 -- PasswordActivity.OnCreate
31/08/2019 19:23:33:314 -- PasswordActivity:apptask=
31/08/2019 19:23:34:86 -- PasswordActivity.OnStart
31/08/2019 19:23:34:111 -- PasswordActivity.OnResume
31/08/2019 19:23:34:112 --  DB null
31/08/2019 19:23:34:117 -- starting: True, Finishing: False, _performingLoad: False
31/08/2019 19:23:34:219 -- PasswordActivity.OnPause
31/08/2019 19:23:34:259 -- Pre-loading database file starting
31/08/2019 19:23:34:316 -- CFS: OpenWhenNoLocalChanges
31/08/2019 19:23:34:318 -- CFS: hashing cached version
31/08/2019 19:23:34:547 -- PasswordActivity.OnResume
31/08/2019 19:23:34:548 --  DB null
31/08/2019 19:23:34:550 -- starting: False, Finishing: False, _performingLoad: False
31/08/2019 19:23:34:698 -- FileSelect.OnStop
31/08/2019 19:23:34:794 -- FileSelect.OnDestroyTrue
31/08/2019 19:23:34:801 -- KeePass.OnStop
31/08/2019 19:23:34:803 -- KeePass.OnDestroyTrue
31/08/2019 19:23:34:805 -- KeePass.OnDestroyTrue
31/08/2019 19:23:36:747 -- CFS: Files in Sync
31/08/2019 19:23:36:766 -- Pre-loading database file completed
31/08/2019 19:23:40:435 -- PasswordActivity.OnPause
31/08/2019 19:23:40:511 -- AppSettingsActivity.OnCreate
31/08/2019 19:23:40:512 -- AppSettingsActivity:apptask=
31/08/2019 19:23:40:847 -- AppSettingsActivity.OnStart
31/08/2019 19:23:40:853 -- AppSettingsActivity.OnResume
31/08/2019 19:23:40:854 --  DB null
31/08/2019 19:23:41:249 -- PasswordActivity.OnStop
```

I was expecting Keepass2Android to use the cached database when there is not an internet connection available. Also, the database file in OneDrive is configured to be available offline.
My Android version is 7.1.2.
Is this a known issue ? Are there any alternatives to keep having the OneDrive synchronization and still can access the database offline ?

Thanks in advance,
José

**Comment by PhilippC** [comment]
will test how this behaves with the new implementation (#55)

**Comment by orerocks** [comment]
I have been having the same problem with Onedrive personal for months on Android 8.0.0. It has worked correctly in the past, falling back to the cached version, but it was a long time ago. Easily reproducible by opening a database (to make sure everything is synced and cached), then closing the database, turning on airplane mode, and then trying to open Keepass2Android. It never gets to the database selection screen.

**Comment by jbujalance** [comment]
@PhilippC, as @orerocks sais, I do think this is a regression, as I also remember this fonctionning a while ago. Also, I use the personal version of OneDrive, not the business one.

**Comment by guitarjorge24** [comment]
I have the same problem with One Drive offline loading. I have Android 9 on a Xiaomi Mi 8. It used to work before but I recently noticed it soenst work anymore. I have tried using both, the latest stable and beta version but neither work without internet.

**Comment by liudve612** [comment]
Same problem for me. Would love to see this fixed in the future!

**Comment by ghanjiboy** [comment]
I am also seeing this issue and would love to see it get fixed, I am on Beta, so can lend and hand as necessary.

**Comment by PhilippC** [comment]
this should be fixed in 1.07b-pre2 (just published to beta). Please click "Change database" (on the password screen), then "Open database", "One Drive", "All files and shared files" (or something else if appropriate) and browse to your file again. This will then use the new implementation. (If you don't re-open the database like this, the old implementation will be used).
Please re-open (or comment) if the issue persists for some reason.
